assets/js/ubi.js

```diff
@@ -0,0 +1,300 @@
+function guiid() {
+  let id = '123456-insecure';
+  try {
+    id = crypto.randomUUID();
+  }
+  catch(error){
+    console.warn('tried to generate a guiid in an insecure context');
+    id ='10000000-1000-4000-8000-100000000000'.replace(/[018]/g, c =>
+    (c ^ crypto.getRandomValues(new Uint8Array(1))[0] & 15 >> c / 4).toString(16));
+  }
+  return id;
+};
+
+export function hash(str, seed=42) {
+  let h1 = 0xdeadbeef ^ seed, h2 = 0x41c6ce57 ^ seed;
+  for(let i = 0, ch; i < str.length; i++) {
+    ch = str.charCodeAt(i);
+    h1 = Math.imul(h1 ^ ch, 2654435761);
+    h2 = Math.imul(h2 ^ ch, 1597334677);
+  }
+  h1  = Math.imul(h1 ^ (h1 >>> 16), 2246822507);
+  h1 ^= Math.imul(h2 ^ (h2 >>> 13), 3266489909);
+  h2  = Math.imul(h2 ^ (h2 >>> 16), 2246822507);
+  h2 ^= Math.imul(h1 ^ (h1 >>> 13), 3266489909);
+
+  return 4294967296 * (2097151 & h2) + (h1 >>> 0);
+};
+
+
+/**
+ * In place of true authentication, this makes a hash out of the user's IP address and browser
+ * for tracking individual user behavior
+ *     user_id = hash( user ip ) + '::' + hash( userAgent )
+ *
+ * NOTE: if this function is called, but user_id starts with 'USER-',
+ *     the function below did not complete successfully,
+ *     and userError() was called instead
+ * @returns
+ */
+export async function initialize(){
+  let i = 1;
+
+  try {
+    if(!sessionStorage.hasOwnProperty('session_id')) {
+      sessionStorage.setItem('session_id', guiid());
+    }
+
+    if(sessionStorage.hasOwnProperty('user_id')){
+      console.log('Already initialized UBI');
+      return;
+    }
+    
+    var rq = new XMLHttpRequest;
+
+    rq.onreadystatechange = function() {
+      if (this.readyState == 4 && this.status == 200) {
+        if(this.response != null) {
+
+          //make a new user id: user ip + '::' + hash( userAgent )
+          let client_id = '';
+          if( window.navigator != null && window.navigator.userAgent != null){
+            client_id = window.navigator.userAgent;
+          } else {
+            client_id = guiid();
+          }
+          let user_id = hash( this.response.ip ) + '::' + hash( client_id );
+          sessionStorage.setItem('user_id', user_id);
+          console.log('user_id: ' + user_id);
+
+        }
+      }
+    };
+
+    rq.onerror = function(){
+      userError();
+      if(this.error != null && this.error != ''){
+        console.error('ERROR Retrieving user info: ' + this.error);
+      }
+      else
+        console.error('UNSPECIFIED ERROR Retrieving user info');
+    }
+
+    rq.open("GET", "https://api64.ipify.org?format=json", true);
```

Comment:
We shouldn't be making calls to external providers. It would be great if we could drop the whole IP-based approach.